benchmarks/benchmarks/common.py

```diff
@@ -25,40 +25,78 @@
     'complex256',
 ]
 
-# values which will be used to construct our sample data matrices
-# replicate 10 times to speed up initial imports of this helper
-# and generate some redundancy
-values = [random.uniform(0, 100) for x in range(nx*ny//10)]*10
-
-squares = {t: numpy.array(values,
-                          dtype=getattr(numpy, t)).reshape((nx, ny))
-           for t in TYPES1}
-
-# adjust complex ones to have non-degenerated imagery part -- use
-# original data transposed for that
-for t, v in squares.items():
-    if t.startswith('complex'):
-        v += v.T*1j
-
-# smaller squares
-squares_ = {t: s[:nxs, :nys] for t, s in squares.items()}
-# vectors
-vectors = {t: s[0] for t, s in squares.items()}
-
-indexes = list(range(nx))
-# so we do not have all items
-indexes.pop(5)
-indexes.pop(95)
-
-indexes_rand = indexes[:]       # copy
-random.shuffle(indexes_rand)         # in-place shuffle
-
-# only now make them arrays
-indexes = numpy.array(indexes)
-indexes_rand = numpy.array(indexes_rand)
-# smaller versions
-indexes_ = indexes[indexes < nxs]
-indexes_rand_ = indexes_rand[indexes_rand < nxs]
+
+class Data(object):
+    """
+    Lazily computed common input data
+    """
+
+    def __getattr__(self, name):
+        if name not in self.__dict__:
```

Comment:
`__getattr__` is only called as a fallback after `__dict__` lookup fails, so the `if` here is redundant. (`__getattribute__` is the real attribute lookup entry point.)

In general the use of `__getattr__` at all seems somewhat over-clever and implicit to me -- wouldn't it be clearer to adjust the benchmark code to call `get_squares` etc. at the appropriate time, so that it's obvious that this is a (possibly memoized) function call? E.g.
```
def memoize_zeroarg_fn(fn):
    value = []
    @functools.wraps(fn)
    def memoized():
        if not value:
            value = [fn()]
        return value[0]

@memoized_zeroarg_fn
def get_squares():
    ...
```

Looks good to me aside from this.